// @crestron/ch5-shell-utilities-cli/build/cli/delete-components/Ch5DeleteComponentsCli.js
"use strict";
// Copyright (C) 2021 to the present, Crestron Electronics, Inc.
// All rights reserved.
// No part of this software may be reproduced in any form, machine
// or natural, without the express written consent of Crestron Electronics.
// Use of this source code is subject to the terms of the Crestron Software License Agreement
// under which you licensed this source code.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ch5DeleteComponentsCli = void 0;
var Ch5BaseClassForCli_1 = require("../Ch5BaseClassForCli");
var Ch5DeleteComponentsCli = /** @class */ (function (_super) {
    __extends(Ch5DeleteComponentsCli, _super);
    /**
     * Constructor
     */
    function Ch5DeleteComponentsCli(showOutputMessages) {
        if (showOutputMessages === void 0) { showOutputMessages = true; }
        var _this = _super.call(this, "delete-components") || this;
        _this.showOutputMessages = showOutputMessages;
        _this.outputResponse = {
            result: false,
            errorMessage: "",
            warningMessage: "",
            data: {
                components: [],
                deleteConfirmation: false
            },
            validInputsForComponentNames: [],
            invalidInputsForComponentNames: []
        };
        _this.pagesAndWidgets = [];
        return _this;
    }
    /**
     * Method for deleting components
     */
    Ch5DeleteComponentsCli.prototype.run = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, 6, 8]);
                        this.checkVersionToExecute();
                        // Initialize
                        return [4 /*yield*/, this.initialize()];
                    case 1:
                        // Initialize
                        _a.sent();
                        // Verify input params
                        return [4 /*yield*/, this.verifyInputParams()];
                    case 2:
                        // Verify input params
                        _a.sent();
                        // Ask details to developer based on input parameter validation
                        return [4 /*yield*/, this.checkPromptQuestions()];
                    case 3:
                        // Ask details to developer based on input parameter validation
                        _a.sent();
                        // Update project-config first (so that if this fails, we don't worry about file deletion). Next Delete Files
                        return [4 /*yield*/, this.processRequest()];
                    case 4:
                        // Update project-config first (so that if this fails, we don't worry about file deletion). Next Delete Files
                        _a.sent();
                        return [3 /*break*/, 8];
                    case 5:
                        e_1 = _a.sent();
                        if (e_1 && this.utils.isValidInput(e_1.message)) {
                            this.outputResponse.errorMessage = e_1.message;
                        }
                        else {
                            this.outputResponse.errorMessage = this.getText("ERRORS.SOMETHING_WENT_WRONG");
                            this.logger.log(e_1);
                        }
                        return [3 /*break*/, 8];
                    case 6: 
                    // Clean up
                    return [4 /*yield*/, this.cleanUp()];
                    case 7:
                        // Clean up
                        _a.sent();
                        return [7 /*endfinally*/];
                    case 8:
                        // Show output response
                        this.logOutput();
                        return [2 /*return*/, this.outputResponse.result]; // The return is required to validate in automation test case
                }
            });
        });
    };
    /**
     * Initialize process
     */
    Ch5DeleteComponentsCli.prototype.initialize = function () {
        this.outputResponse = {
            result: false,
            errorMessage: "",
            warningMessage: "",
            data: {
                components: [],
                deleteConfirmation: false
            },
            validInputsForComponentNames: [],
            invalidInputsForComponentNames: []
        };
        this.pagesAndWidgets = this.projectConfig.getAllPagesAndWidgets();
    };
    /**
     * Verify input parameters
     */
    Ch5DeleteComponentsCli.prototype.verifyInputParams = function () {
        if (this.pagesAndWidgets.length === 0) {
            throw new Error(this.getText("ERRORS.NO_PAGES_WIDGETS_AVAILABLE_IN_PROJECT"));
        }
        var listOfInputComponents = this.inputArguments["list"];
        if (listOfInputComponents && listOfInputComponents.length > 0) {
            var _loop_1 = function (i) {
                var componentObject = this_1.pagesAndWidgets.find(function (tempObj) { return tempObj.name.trim().toLowerCase() === listOfInputComponents[i].trim().toLowerCase(); });
                if (componentObject) {
                    this_1.outputResponse.validInputsForComponentNames.push(listOfInputComponents[i]);
                }
                else {
                    if (this_1.utils.isValidInput(listOfInputComponents[i])) {
                        // This code is to check if empty input values are provided.
                        this_1.outputResponse.invalidInputsForComponentNames.push(listOfInputComponents[i]);
                    }
                }
            };
            var this_1 = this;
            for (var i = 0; i < listOfInputComponents.length; i++) {
                _loop_1(i);
            }
            if (this.outputResponse.validInputsForComponentNames.length === 0) {
                this.outputResponse.warningMessage = this.getText("ERRORS.INVALID_PARAM_INPUTS");
            }
            else {
                this.outputResponse.data.components = this.outputResponse.validInputsForComponentNames;
                if (this.outputResponse.invalidInputsForComponentNames.length > 0) {
                    for (var i = 0; i < this.outputResponse.invalidInputsForComponentNames.length; i++) {
                        this.outputResponse.warningMessage += this.outputResponse.invalidInputsForComponentNames[i] + "\n";
                    }
                }
            }
        }
        if (this.utils.isValidInput(this.outputResponse.warningMessage)) {
            if (this.showOutputMessages === true) {
                this.logger.printWarning(this.getText("ERRORS.MESSAGE_TITLE", this.outputResponse.warningMessage));
            }
        }
    };
    /**
     * Check if there are questions to be prompted to the developer
     */
    Ch5DeleteComponentsCli.prototype.checkPromptQuestions = function () {
        return __awaiter(this, void 0, void 0, function () {
            var choicesList, i, componentType, componentsQuery, _a, questionsArray, _b;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!(this.outputResponse.data.components.length === 0)) return [3 /*break*/, 2];
                        choicesList = [];
                        for (i = 0; i < this.pagesAndWidgets.length; i++) {
                            componentType = (this.pagesAndWidgets[i].type === "page") ? "Page" : "Widget";
                            choicesList.push({ value: i, hint: this.getText("HINT_COMPONENT_DETAILS", componentType, this.pagesAndWidgets[i].component.fullPath + this.pagesAndWidgets[i].component.fileName), name: this.pagesAndWidgets[i].name, component: this.pagesAndWidgets[i].component, type: this.pagesAndWidgets[i].type });
                        }
                        this.logger.log("choicesList", choicesList);
                        componentsQuery = new this.getMultiSelect({
                            name: 'value',
                            message: this.getText("VALIDATIONS.SELECT_COMPONENT_TO_DELETE"),
                            choices: choicesList
                        });
                        _a = this.outputResponse.data;
                        return [4 /*yield*/, componentsQuery.run()
                                .then(function (selectedComponents) { return selectedComponents; })
                                .catch(function (error) { return []; })];
                    case 1:
                        _a.components = _c.sent();
                        _c.label = 2;
                    case 2:
                        this.logger.log("Components selected", this.outputResponse.data.components);
                        if (!(this.outputResponse.data.components.length > 0)) return [3 /*break*/, 6];
                        this.outputResponse.validInputsForComponentNames = this.outputResponse.data.components;
                        if (!(this.inputArguments["force"] === true)) return [3 /*break*/, 3];
                        this.outputResponse.data.deleteConfirmation = true;
                        return [3 /*break*/, 5];
                    case 3:
                        questionsArray = [
                            {
                                type: 'select',
                                name: 'deleteConfirmation',
                                message: this.getText("VALIDATIONS.ARE_YOU_SURE_TO_DELETE"),
                                choices: [
                                    { message: this.getText("VALIDATIONS.CONFIRMATION_OPTION_YES"), hint: this.getText("VALIDATIONS.CONFIRMATION_OPTION_YES_DESCRIPTION"), value: "Y" },
                                    { message: this.getText("VALIDATIONS.CONFIRMATION_OPTION_NO"), hint: this.getText("VALIDATIONS.CONFIRMATION_OPTION_NO_DESCRIPTION"), value: "N" }
                                ],
                                initial: 0
                            }
                        ];
                        _b = this.outputResponse.data;
                        return [4 /*yield*/, this.getEnquirer.prompt(questionsArray)
                                .then(function (response) {
                                _this.logger.log(response);
                                return _this.utils.convertStringToBoolean(response.deleteConfirmation);
                            })
                                .catch(function (err) {
                                return false;
                            })];
                    case 4:
                        _b.deleteConfirmation = _c.sent();
                        _c.label = 5;
                    case 5: return [3 /*break*/, 7];
                    case 6: throw new Error(this.getText("ERRORS.NO_COMPONENTS_FOR_DELETION"));
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Implement this component's main purpose
     */
    Ch5DeleteComponentsCli.prototype.processRequest = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _loop_2, this_2, i;
            var _this = this;
            return __generator(this, function (_a) {
                if (this.outputResponse.data.deleteConfirmation === true) {
                    _loop_2 = function (i) {
                        var componentObject = this_2.pagesAndWidgets.find(function (tempObj) { return tempObj.name.trim().toLowerCase() === _this.outputResponse.data.components[i].trim().toLowerCase(); });
                        if (componentObject) {
                            this_2.utils.deleteFolder(componentObject.component.fullPath);
                        }
                        else {
                            throw new Error(this_2.getText("ERRORS.SOMETHING_WENT_WRONG"));
                        }
                    };
                    this_2 = this;
                    for (i = 0; i < this.outputResponse.data.components.length; i++) {
                        _loop_2(i);
                    }
                    // Change project config
                    this.projectConfig.removePagesFromJSON(this.outputResponse.data.components);
                    this.projectConfig.removeWidgetsFromJSON(this.outputResponse.data.components);
                    this.outputResponse.result = true;
                }
                else if (this.outputResponse.data.deleteConfirmation === false) {
                    throw new Error(this.getText("ERRORS.DO_NOT_DELETE_COMPONENTS"));
                }
                else {
                    throw new Error(this.getText("ERRORS.PROGRAM_STOPPED_OR_UNKNOWN_ERROR"));
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Clean up
     */
    Ch5DeleteComponentsCli.prototype.cleanUp = function () {
        // Nothing to cleanup for this process
    };
    /**
     * Log Final Response Message
     */
    Ch5DeleteComponentsCli.prototype.logOutput = function () {
        if (this.showOutputMessages === true) {
            if (this.outputResponse.result === false) {
                this.logger.printError(this.outputResponse.errorMessage);
            }
            else {
                this.logger.printSuccess(this.getText("SUCCESS_MESSAGE", this.utils.convertArrayToString(this.outputResponse['validInputsForComponentNames'], ", ")) + "\n");
                if (this.outputResponse['invalidInputsForComponentNames'].length > 0) {
                    this.logger.printSuccess(this.getText("SUCCESS_MESSAGE_WITH_EXCEPTION", this.utils.convertArrayToString(this.outputResponse['invalidInputsForComponentNames'], ", ")) + "\n");
                }
                this.logger.printSuccess(this.getText("SUCCESS_MESSAGE_CONCLUSION", this.utils.convertArrayToString(this.outputResponse['validInputsForComponentNames'], ", ")) + "\n");
            }
        }
    };
    return Ch5DeleteComponentsCli;
}(Ch5BaseClassForCli_1.Ch5BaseClassForCli));
exports.Ch5DeleteComponentsCli = Ch5DeleteComponentsCli;
